docs/source/getting-started/testing.rst

```diff
@@ -0,0 +1,105 @@
+The first tool is a general JavaScript testing framework called
+`Jest <https://jestjs.io/>`_.
+Jest allows you to write test cases containing expectations about behaviour should occur.
+If the expectations for a test are not met, then you should see helpful messages about what went wrong.
+You can also write setup and cleanup procedures that run before and after each test.
+This will be very useful for us to manage services like the DAML sandbox and JSON API server.
+
+The second tool we use heavily is a library called
+`Puppeteer <https://pptr.dev/>`_.
+Puppeteer allows you to control a Chrome browser from a NodeJS program.
+In our case we use it to perform actions like logging in, following friends and sending messages.
+We can also control multiple browser pages at once and test the interactions between different users.
+As a bonus, if something goes wrong you can take a screenshot and see what state your test browser got into!
+
+Let's see how to use these tools to write some tests for our social network app.
+You can see the full suite in the file ``index.test.tsx``.
+We first have some global state that we will use throughout our tests.
+Specifically, we have child processes for the ``daml start`` and ``yarn start`` commands, which run for the duration of our tests.
+We also have a single Puppeteer browser that we share among tests, opening new pages in each.
+
+The ``beforeAll()`` section is a function run once before any of the tests run.
+We use it to spawn the ``daml start`` and ``yarn start`` processes and launch the browser.
+On the other hand the ``afterAll()`` section is used to shut down these processes and close the browser.
+This step is important to prevent child processes persisting in the background after our program has finished.
+
+Now let's get to the tests!
+The idea is to control the browser in the same way we would expect a user to in each scenario we want to test.
+This means we use Puppeteer to type text into input forms, click buttons and search for particular elements on the page.
+In order to find those elements, we do need to make some adjustments in our React components, which we'll show later.
+Let's start at a higher level with a ``test``.
+
+.. literalinclude:: code/ui-before/index.test.tsx
+  :language: tsx
+  :start-after: // BEGIN_LOGIN_TEST
+  :end-before: // END_LOGIN_TEST
+
+We'll walk though this step by step.
+
+    - The ``test`` syntax is provided by Jest to indicate a new test running the function given as an argument (along with a description and time limit).
+    - ``getParty()`` gives us a new party name. Right now it is just a string unique to this set of tests, but in the future we will use the Party Management Service to allocate parties.
+    - ``newUiPage()`` is a helper function that uses the Puppeteer browser to open a new page (we use one page per party in these tests), navigate to the app URL and return a ``Page`` object.
```

Comment:
These are the basics for any test, right?

Reply:
That's right, I'll try to mention that.